src/category_theory/connected.lean

```diff
@@ -0,0 +1,336 @@
+list.exists_zigzag (connected_zigzag _ _)
+
+/--
+If any two objects in an inhabited category are linked by a sequence of (potentially reversed)
+morphisms, then J is connected.
+
+The converse of `exists_zigzag'`.
+-/
+def connected_of_zigzag [inhabited J]
+  (h : ∀ (j₁ j₂ : J), ∃ (l : list J), list.chain' zag l ∧ ∃ (hl : l ≠ []), list.head_of_ne_nil l hl = j₁ ∧ list.last l hl = j₂) :
+  connected J :=
+begin
+  apply connected.of_induct,
+  intros p d k j,
+  obtain ⟨l, zags, nemp, hd, tl⟩ := h j (default J),
+  apply list.prop_up_chain' p l nemp zags hd tl _ d,
+  rintros _ _ (⟨⟨_⟩⟩ | ⟨⟨_⟩⟩),
+  { exact (k a).2 },
+  { exact (k a).1 }
+end
+
+end J
+
+section examples
+instance cospan_inhabited : inhabited walking_cospan := ⟨walking_cospan.one⟩
+
+instance cospan_connected : connected (walking_cospan) :=
+begin
+  apply connected.of_induct,
+  introv _ t,
+  cases j,
+  { rwa t walking_cospan.hom.inl },
+  { rwa t walking_cospan.hom.inr },
+  { assumption }
+end
+
+instance span_inhabited : inhabited walking_span := ⟨walking_span.zero⟩
+
+instance span_connected : connected (walking_span) :=
+begin
+  apply connected.of_induct,
+  introv _ t,
+  cases j,
+  { assumption },
+  { rwa ← t walking_span.hom.fst },
+  { rwa ← t walking_span.hom.snd },
+end
+
+instance parallel_pair_inhabited : inhabited walking_parallel_pair := ⟨walking_parallel_pair.one⟩
+
+instance parallel_pair_connected : connected (walking_parallel_pair) :=
+begin
+  apply connected.of_induct,
+  introv _ t, cases j,
+  { rwa t walking_parallel_pair_hom.left },
+  { assumption }
+end
+
+end examples
+
+section C
+variables {J : Type v₂} [𝒥 : category.{v₁} J]
+include 𝒥
+
+variables {C : Type u₂} [𝒞 : category.{v₂} C]
+include 𝒞
+
+/--
+For objects `X Y : C`, any natural transformation `α : const X ⟶ const Y` from a connected
+category must be constant.
+This is the key property of connected categories which we use to establish properties about limits.
+-/
+lemma nat_trans_from_connected [conn : connected J] {X Y : C}
+  (α : (functor.const J).obj X ⟶ (functor.const J).obj Y) :
+  ∀ (j : J), α.app j = (α.app (default J) : X ⟶ Y) :=
+@constant_of_preserves_morphisms _ _ _
+  (X ⟶ Y)
+  (λ j, α.app j)
+  (λ _ _ f, (by { have := α.naturality f, erw [id_comp, comp_id] at this, exact this.symm }))
+
+end C
+
+local attribute [tidy] tactic.case_bash
+
+variables {C : Type u₂} [𝒞 : category.{v₂} C]
+include 𝒞
+
+section products
+
+variables [has_binary_products.{v₂} C]
+
+variables {J : Type v₂} [small_category J]
+
+/-- (Impl). The obvious natural transformation from (X × K -) to K. -/
+@[simps]
+def γ₂ {K : J ⥤ C} (X : C) : K ⋙ prod_functor.obj X ⟶ K :=
+{ app := λ Y, limits.prod.snd }
+
+/-- (Impl). The obvious natural transformation from (X × K -) to X -/
+@[simps]
+def γ₁ {K : J ⥤ C} (X : C) : K ⋙ prod_functor.obj X ⟶ (functor.const J).obj X :=
+{ app := λ Y, limits.prod.fst }
```

Comment:
Perhaps stick these inside a `prod_preserves_connected_limits` namespace to make it clearer that these are private implementation details (and to get out of the obligation of finding good names for them).